### PR TITLE
Add YttmTokenizer, ImageTextDataset from @rom1504, Single-GPU trainin…

### PR DESCRIPTION
**Pull request by afiaka87**
…g script

### CODE REVIEW

File: x_clip/dataset.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import torch
+from pathlib import Path
+from PIL import UnidentifiedImageError, Image
+from torch.utils.data import Dataset
+
+
+class ImageTextDataset(Dataset):
```
Comment (rom1504) [inline x_clip/dataset.py:7]: I think there might be some value to making an independent package with such vision/text datasets (this one and webdataset at least), and depending on it here and in dalle pytorch (and in clip retrieval and probably a bunch of other places)
what do you think?

Comment (afiaka87) [reply x_clip/dataset.py:7]: agreed. at this point it's a bit strange that pytorch doesn't have something for this by now - are you aware of anything?

Comment (rom1504) [reply x_clip/dataset.py:7]: besides from discussions at https://github.com/pytorch/pytorch/issues/38419 no I don't know
I think pytorch has been taking the approach of letting the user build their own things on the data side

although as a reminder image+text dataset is something that has begun being useful in 2021, it's not that old yet :)

File: train.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    # Create optimizer
+    optimizer = torch.optim.Adam(clip_model.parameters(), lr=args.learning_rate)
+
+    # Train
+    for epoch in tqdm.tqdm(range(args.num_epochs)):
+        current_epoch_pbar = tqdm.tqdm(enumerate(image_text_dataloader), total=len(image_text_dataloader), unit='batch', unit_scale=args.batch_size)
+        for batch_idx, (text, images) in current_epoch_pbar:
+            with autocast(enabled=args.amp):
+                text, images = map(lambda t: t.cuda(), (text, images))
+                mask = torch.ones_like(text).bool()
```
Comment (MicPie) [inline train.py:117]: A recent update adapted the text loss to incorporate a text mask: https://github.com/lucidrains/x-clip/blob/ac62779db5954802e55ff97f03df37f1985d8d4f/x_clip/x_clip.py#L314.
To make use of this, the dataset could return the accompanying text masks to utilize the masked mean.

**Comment by afiaka87** [comment]
@lucidrains Let me know if there's any glaring mistakes but this should provide similar functionality to what we had in dalle-pytorch. Main thing missing is webdataset support, but I figured folks may want to start using this and I don't know how long it will take me to implement that. 

Romain made a decent point about how everyone seems to just rewrite/copy-paste the text-image dataloader but unfortunately I can't commit to maintaining a pip package for that either.

**Comment by afiaka87** [comment]
@MicPie Thanks for the DDP code. I've rebased your branch onto this one so we can hopefully get that upstream.

**Comment by afiaka87** [comment]
Apologies, have not had the time to get this branch working. Closing for now.
